e2e/nomostest/sync.go

```diff
@@ -276,47 +281,49 @@ func resourceGroupHasReconciled(sourceHash string) testpredicates.Predicate {
 			return testpredicates.WrongTypeErr(o, &v1alpha1.ResourceGroup{})
 		}
 
-		// Status is disabled for this ResourceGroup, skip all status checks
+		// If status is disabled for this ResourceGroup, skip all status checks
 		if resourcegroup.IsStatusDisabled(rg) {
 			return testpredicates.ResourceGroupHasNoStatus()(o)
 		}
 
-		if len(rg.Spec.Resources) != len(rg.Status.ResourceStatuses) {
-			return fmt.Errorf("length of spec.resources (%d) does not equal length of status.resourceStatuses (%d)",
-				len(rg.Spec.Resources), len(rg.Status.ResourceStatuses))
+		// Test status.observedGeneration matches metadata.generation
+		if err := testpredicates.HasObservedLatestGeneration(scheme)(o); err != nil {
+			return err
 		}
 
-		resourceCount := make(map[v1alpha1.ObjMetadata]int)
-		for _, s := range rg.Status.ResourceStatuses {
-			if err := resourceStatusIsCurrent(s, sourceHash); err != nil {
-				return fmt.Errorf("object %s is not current: %w", s.ObjMetadata, err)
-			}
-			resourceCount[s.ObjMetadata]++
+		// Test that metadata.deletionTimestamp is missing, and conditions do not include Reconciling=True or Stalled=True
+		if err := testpredicates.StatusEquals(scheme, kstatus.CurrentStatus)(o); err != nil {
+			return err
+		}
+
+		// Test that Stalled condition exists and is "False".
+		// This ensures the condition wasn't removed by the applier.
+		if err := testpredicates.HasConditionStatus(scheme, string(v1alpha1.Stalled), corev1.ConditionFalse)(o); err != nil {
+			return err
 		}
 
-		for _, o := range rg.Spec.Resources {
-			if count, ok := resourceCount[o]; ok && count > 0 {
-				resourceCount[o]--
-			} else {
-				return fmt.Errorf("spec.resources does not equal status.resourceStatuses")
+		// Test that all the spec.resources exist in the status.resourceStatuses,
+		// and that all deleted objects were removed by the applier.
+		shortCommit := resourcegroup.TruncateSourceHash(commit)
+		found := rg.Status.ResourceStatuses
+		expected := make([]v1alpha1.ResourceStatus, len(rg.Spec.Resources))
+		for i, resource := range rg.Spec.Resources {
+			expected[i] = v1alpha1.ResourceStatus{
+				ObjMetadata: *resource.DeepCopy(),
+				Status:      v1alpha1.Current,
+				Strategy:    v1alpha1.Apply,
+				Actuation:   v1alpha1.ActuationSucceeded,
+				Reconcile:   v1alpha1.ReconcileSucceeded,
+				SourceHash:  shortCommit,
+				Conditions:  nil,
 			}
 		}
-		return nil
-	}
-}
+		if !equality.Semantic.DeepEqual(found, expected) {
```

Comment:
I'm actually pretty surprised this passed all the e2e tests... I thought I was going to have to sort them.

But maybe we don't have any tests that churn in the right way to break the ordering...

Reply:
Well, I guess if it breaks we can always add sorting later.